File: src/main/java/com/heekng/celloct/controller/StaffWorkTimeController.java
package com.heekng.celloct.controller;

import com.heekng.celloct.config.oauth.dto.SessionMember;
import com.heekng.celloct.dto.ShopDto;
import com.heekng.celloct.dto.WorkDto;
import com.heekng.celloct.dto.WorkUpdateRequestDto;
import com.heekng.celloct.entity.Shop;
import com.heekng.celloct.entity.Staff;
import com.heekng.celloct.entity.Work;
import com.heekng.celloct.repository.StaffRepository;
import com.heekng.celloct.repository.WorkRepository;
import com.heekng.celloct.service.ShopService;
import com.heekng.celloct.service.WorkService;
import com.heekng.celloct.service.WorkUpdateRequestService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Controller;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.*;

import javax.servlet.http.HttpSession;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Controller
@Slf4j
@RequiredArgsConstructor
@RequestMapping("/staff/{shopId}")
public class StaffWorkTimeController {

    private final HttpSession httpSession;
    private final StaffRepository staffRepository;
    private final ShopService shopService;
    private final WorkRepository workRepository;
    private final WorkService workService;
    private final WorkUpdateRequestService workUpdateRequestService;

    @GetMapping("/workTime/insert")
    public String insertWorkTime(@PathVariable("shopId") Long shopId, Model model) {
        SessionMember member = (SessionMember) httpSession.getAttribute("member");
        Optional<Staff> staffOptional = staffRepository.findByMemberIdAndShopId(member.getId(), shopId);
        if (staffOptional.isEmpty()) {
            return "redirect:/";
        }

        Shop shop = shopService.findShop(shopId);
        model.addAttribute("shop", new ShopDto.ShopDetailResponse(shop));

        return "staff/workTimeInsert";
    }

    @GetMapping("/workTime/validateExist")
    @ResponseBody
    public Boolean checkExist(@PathVariable("shopId") Long shopId, WorkDto.CheckExistRequest checkExistRequest) {
        SessionMember member = (SessionMember) httpSession.getAttribute("member");
        Optional<Staff> staffOptional = staffRepository.findByMemberIdAndShopId(member.getId(), shopId);
        if (staffOptional.isEmpty()) {
            return false;
        }

        checkExistRequest.setMemberId(member.getId());
        checkExistRequest.setShopId(shopId);
        return workService.validateExist(checkExistRequest);
    }

    @PostMapping("/workTime/insert")
    public String doInsertWorkTime(@PathVariable("shopId") Long shopId, WorkDto.AddRequest addRequest, Model model) {
        SessionMember member = (SessionMember) httpSession.getAttribute("member");
        Optional<Staff> staffOptional = staffRepository.findByMemberIdAndShopId(member.getId(), shopId);
        if (staffOptional.isEmpty()) {
            return "redirect:/";
        }

        addRequest.setMemberId(member.getId());
        addRequest.setShopId(shopId);
        Long workId = workService.addWork(addRequest);

        return "redirect:/";
    }

    @GetMapping("/workTimes")
    public String workTimes(@PathVariable("shopId") Long shopId, Model model) {
        SessionMember member = (SessionMember) httpSession.getAttribute("member");
        Optional<Staff> staffOptional = staffRepository.findByMemberIdAndShopId(member.getId(), shopId);
        if (staffOptional.isEmpty()) {
            return "redirect:/";
        }

        Shop shop = shopService.findShop(shopId);
        model.addAttribute("shop", new ShopDto.ShopDetailResponse(shop));

        return "staff/workTime";
    }

    @GetMapping("/workTimes/search")
    @ResponseBody
    public List<WorkDto.WorkResponse> findWorks(
            @PathVariable("shopId") Long shopId,
            @RequestParam("year") Integer year,
            @RequestParam("month") Integer month
    ) {
        SessionMember member = (SessionMember) httpSession.getAttribute("member");
        Optional<Staff> staffOptional = staffRepository.findByMemberIdAndShopId(member.getId(), shopId);
        if (staffOptional.isEmpty()) {
            throw new IllegalStateException("해당 매장의 직원이 아닙니다.");
        }

        Staff staff = staffRepository.findByMemberIdAndShopId(member.getId(), shopId).orElseThrow(() -> new IllegalStateException("존재하지 않는 직원입니다."));

        WorkDto.FindWorkRequest findWorkRequest = WorkDto.FindWorkRequest.builder()
                .staffId(staff.getId())
                .shopId(shopId)
                .year(year)
                .month(month)
                .build();
        return workService.findWorkByFindWorkRequest(findWorkRequest).stream()
                .map(WorkDto.WorkResponse::new)
                .collect(Collectors.toList());
    }

    @GetMapping("/workTimes/{workId}")
    @ResponseBody
    public WorkDto.WorkDetailResponse workDetail(
            @PathVariable("shopId") Long shopId,
            @PathVariable("workId") Long workId
    ) {
        SessionMember member = (SessionMember) httpSession.getAttribute("member");
        Optional<Staff> staffOptional = staffRepository.findByMemberIdAndShopId(member.getId(), shopId);
        if (staffOptional.isEmpty()) {
            throw new IllegalStateException("해당 매장의 직원이 아닙니다.");
        }

        Staff staff = staffRepository.findByMemberIdAndShopId(member.getId(), shopId).orElseThrow(() -> new IllegalStateException("해당 직원의 근무가 아닙니다."));
        Work work = workRepository.findByIdAndStaffId(workId, staff.getId()).orElseThrow(() -> new IllegalStateException("존재하지 않는 근무입니다"));
        return new WorkDto.WorkDetailResponse(work);
    }

    @Transactional
    @PostMapping("/workTimes/{workId}/update")
    @ResponseBody
    public Boolean updateRequest(
            @PathVariable("shopId") Long shopId,
            @PathVariable("workId") Long workId,
            @RequestBody WorkUpdateRequestDto.AddRequest addRequest
    ) {
        SessionMember member = (SessionMember) httpSession.getAttribute("member");
        Optional<Staff> staffOptional = staffRepository.findByMemberIdAndShopId(member.getId(), shopId);
        if (staffOptional.isEmpty()) {
            throw new IllegalStateException("해당 매장의 직원이 아닙니다.");
        }

        Staff staff = staffRepository.findByMemberIdAndShopId(member.getId(), shopId).orElseThrow(() -> new IllegalStateException("해당 매장의 직원이 아닙니다."));

        addRequest.setWorkId(workId);
        addRequest.setStaffId(staff.getId());
        workUpdateRequestService.addWorkUpdateRequest(addRequest);
        return true;
    }
}

File: src/main/kotlin/com/heekng/celloct/controller/StaffWorkTimeController.kt
package com.heekng.celloct.controller

import com.heekng.celloct.annotation.RoleCheck
import com.heekng.celloct.annotation.enums.UserType
import com.heekng.celloct.config.oauth.dto.SessionMember
import com.heekng.celloct.dto.ShopDto
import com.heekng.celloct.dto.WorkDto
import com.heekng.celloct.dto.WorkUpdateRequestDto
import com.heekng.celloct.repository.ShopRepository
import com.heekng.celloct.repository.StaffRepository
import com.heekng.celloct.repository.WorkRepository
import com.heekng.celloct.service.WorkService
import com.heekng.celloct.service.WorkUpdateRequestService
import com.heekng.celloct.util.findByIdOrThrow
import org.springframework.stereotype.Controller
import org.springframework.transaction.annotation.Transactional
import org.springframework.ui.Model
import org.springframework.web.bind.annotation.*
import javax.servlet.http.HttpSession

@Controller
@RequestMapping("/staff/{shopId}")
class StaffWorkTimeController(
    private val httpSession: HttpSession,
    private val staffRepository: StaffRepository,
    private val shopRepository: ShopRepository,
    private val workRepository: WorkRepository,
    private val workService: WorkService,
    private val workUpdateRequestService: WorkUpdateRequestService,
) {

    @GetMapping("/workTime/insert")
    @RoleCheck(UserType.STAFF)
    fun insertWorkTime(
        @PathVariable("shopId") shopId: Long,
        model: Model,
    ): String {
        val shop = shopRepository.findByIdOrThrow(shopId)
        model.addAttribute("shop", ShopDto.ShopDetailResponse(shop))

        return "staff/workTimeInsert"
    }

    @GetMapping("/workTime/validateExist")
    @ResponseBody
    @RoleCheck(UserType.STAFF, true)
    fun checkExist(
        @PathVariable("shopId") shopId: Long,
        checkExistRequest: WorkDto.CheckExistRequest,
    ): Boolean {
        val sessionMember = httpSession.getAttribute("member") as SessionMember?

        checkExistRequest.memberId = sessionMember!!.id
        checkExistRequest.shopId = shopId

        return workService.validateExist(checkExistRequest)
    }

    @PostMapping("/workTime/insert")
    @RoleCheck(UserType.STAFF)
    fun doInsertWorkTime(
        @PathVariable("shopId") shopId: Long,
        addRequest: WorkDto.AddRequest,
    ): String {
        val sessionMember = httpSession.getAttribute("member") as SessionMember?

        addRequest.memberId = sessionMember!!.id
        addRequest.shopId = shopId
        workService.addWork(addRequest)
        return "redirect:/"
    }

    @GetMapping("/workTimes")
    @RoleCheck(UserType.STAFF)
    fun workTimes(
        @PathVariable("shopId") shopId: Long,
        model: Model,
    ): String {
        val shop = shopRepository.findByIdOrThrow(shopId)
        model.addAttribute("shop", ShopDto.ShopDetailResponse(shop))

        return "staff/workTime"
    }

    @GetMapping("/workTimes/search")
    @ResponseBody
    @RoleCheck(UserType.STAFF, true)
    fun findWorks(
        @PathVariable("shopId") shopId: Long,
        @RequestParam("year") year: Int,
        @RequestParam("month") month: Int,
    ): List<WorkDto.WorkResponse> {
        val sessionMember = httpSession.getAttribute("member") as SessionMember?
        val staff = staffRepository.findOneQ(
            memberId = sessionMember!!.id,
            shopId = shopId,
        )

        val findWorkRequest = WorkDto.FindWorkRequest(
            staffId = staff!!.id!!,
            shopId = shopId,
            year = year,
            month = month,
        )
        return workService.findWorkByFindWorkRequest(findWorkRequest)
            .map { work -> WorkDto.WorkResponse(work) }
    }

    @GetMapping("/workTimes/{workId}")
    @ResponseBody
    @RoleCheck(UserType.STAFF, true)
    fun workDetail(
        @PathVariable("shopId") shopId: Long,
        @PathVariable("workId") workId: Long,
    ): WorkDto.WorkDetailResponse {
        val sessionMember = httpSession.getAttribute("member") as SessionMember?
        val staff = staffRepository.findOneQ(
            memberId = sessionMember!!.id,
            shopId = shopId,
        )

        val work = workRepository.findByIdAndStaffId(workId, staff!!.id!!)
            ?: throw IllegalStateException("존재하지 않는 근무입니다.")
        return WorkDto.WorkDetailResponse(work)
    }

    @Transactional
    @PostMapping("/workTimes/{workId}/update")
    @ResponseBody
    @RoleCheck(UserType.STAFF, true)
    fun updateRequest(
        @PathVariable("shopId") shopId: Long,
        @PathVariable("workId") workId: Long,
        @RequestBody addRequest: WorkUpdateRequestDto.AddRequest,
    ): Boolean {
        val sessionMember = httpSession.getAttribute("member") as SessionMember?
        val staff = staffRepository.findOneQ(
            memberId = sessionMember!!.id,
            shopId = shopId
        )

        addRequest.workId = workId
        addRequest.staffId = staff!!.id!!
        workUpdateRequestService.addWorkUpdateRequest(addRequest)
        return true
    }

}